elasticsearch/src/main/scala/com/stratio/crossdata/connector/elasticsearch/ElasticSearchRowConverter.scala

```diff
@@ -18,75 +18,87 @@ package com.stratio.crossdata.connector.elasticsearch
-    import scala.collection.JavaConverters._
-    val schemaMap = schema.map(field => field.name -> field.dataType).toMap
 
     array map { hit =>
-      hitAsRow(hit.fields().asScala.toMap, schemaMap, requiredFields.map(_.name))
+      hitAsRow(
+        hit.fields().asScala.toMap,
+        Option(hit.getSource).map(_.asScala.toMap).getOrElse(Map.empty),
+        schema,
+        requiredFields.map(_.name)
+      )
     }
   }
 
   def hitAsRow(
                 hitFields: Map[String, SearchHitField],
-                schemaMap: Map[String, DataType],
+                subDocuments: Map[String, AnyRef],
+                schema: StructType,
                 requiredFields: Seq[String]): Row = {
-    val values: Seq[Any] = requiredFields.map {
-      name =>
-        hitFields.get(name).flatMap(v => Option(v)).map(
-          toSQL(_, schemaMap(name))).orNull
-    }
-    Row.fromSeq(values)
-  }
 
-  def toSQL(value: SearchHitField, dataType: DataType): Any = {
+    val schemaMap = schema.map(field => field.name -> field.dataType).toMap
+
+    val values: Seq[Any] = requiredFields.map { name =>
 
-    Option(value).map { case value =>
-      //Assure value is mapped to schema constrained type.
-      enforceCorrectType(value.getValue, dataType)
-    }.orNull
-  }
+      // TODO: What if a nested subdocument is targeted
+      (hitFields.get(name) orElse subDocuments.get(name)).flatMap(Option(_)) map {
+        ((value: Any) => enforceCorrectType(value, schemaMap(name))) compose {
+          case hitField: SearchHitField => hitField.getValue
+          case other => other
+        }
+      } orNull
 
+    }
+    new GenericRowWithSchema(values.toArray, schema)
+  }
 
-  protected def enforceCorrectType(value: Any, desiredType: DataType): Any = {
+  protected def enforceCorrectType(value: Any, desiredType: DataType): Any =
       // TODO check if value==null
-      Option(desiredType).map {
+      desiredType match {
         case StringType => value.toString
         case _ if value == "" => null // guard the non string type
+        case ByteType => toByte(value)
+        case ShortType => toShort(value)
         case IntegerType => toInt(value)
         case LongType => toLong(value)
         case DoubleType => toDouble(value)
+        case FloatType => toFloat(value)
         case DecimalType() => toDecimal(value)
         case BooleanType => value.asInstanceOf[Boolean]
         case TimestampType => toTimestamp(value)
         case NullType => null
         case DateType => toDate(value)
+        case BinaryType => toBinary(value)
+        case schema: StructType => toRow(value, schema)
         case _ =>
           sys.error(s"Unsupported datatype conversion [${value.getClass}},$desiredType]")
           value
-      }.orNull
+      }
+
+  private def toByte(value: Any): Byte = value match {
+    case value: Byte => value
+    case value: Int => value.toByte
+    case value: Long => value.toByte
+  }
```

Comment:
I think it should crash when no available conversions are found. The default case would only make sense if it logged the error and raised an exception. I didn't want to increase the set of changes in the PR to change  `MatchError` for any other exception. I would do it at a second PR.